# Intro: opening file and checking data

First, we import the libraries that we will need to use. By convention, we leave them all on the beginning of the script. If later we realize we need other libraries, we will come back and put them here.

In [1]:
import pandas as pd

Pandas `read_csv()` function grabs the values from a csv file and transforms it on a dataframe object. 

In [2]:
irs = pd.read_csv('bmf.bm1812.csv')

/home/rafa/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


The error message means we are not doing the most efficient way of making this conversion, but we shouldn't be concerned with that.



We use some basic functions to take a quick look at our dataframe. 
`irs.shape` returns the number of rows and columns. In this case, both are huge.
`irs.head()` shows only the first 5 observations. Jupyter notebook shows only a limited number of columns, which depend on the screen size. 
To see a list with all the columns, we use `irs.columns`


In [3]:
irs.shape

(1499450, 40)

In [4]:
irs.head()

EIN SEC_NAME  FRCD  SUBSECCD  TAXPER  ASSETS  INCOME  \
0   19818     3514    60         3     NaN     NaN     NaN   
1   29215      NaN    60         3     NaN     NaN     NaN   
2  260049      NaN    60         3     NaN     NaN     NaN   
3  490336      NaN    60         3     NaN     NaN     NaN   
4  587764      NaN    60         3     NaN     NaN     NaN   

                           NAME            ADDRESS          CITY   ...    \
0  PALMER SECOND BAPTIST CHURCH  1050 THORNDIKE ST        PALMER   ...     
1           ST GEORGE CATHEDRAL     523 E BROADWAY  SOUTH BOSTON   ...     
2        CORINTH BAPTIST CHURCH          PO BOX 92       HOSFORD   ...     
3       EASTSIDE BAPTIST CHURCH         PO BOX 296       LABELLE   ...     
4          IGLESIA BETHESDA INC     157 ANDOVER ST        LOWELL   ...     

  LEVEL4 LEVEL1 NTMAJ10  MAJGRPB  LEVEL3 LEVEL2 NTMAJ12  NTMAJ5  FILER  ZFILER  
0      X     PC      RE        X      RE      O      RE      OT      N       N  
1      X     PC      RE        X      RE      O      RE      OT      N       N  
2      X     PC      RE        X      RE      O      RE      OT      N       N  
3      X     PC      RE        X      RE      O      RE      OT      N       N  
4      X     PC      RE        X      RE      O      RE      OT      N       N  

[5 rows x 40 columns]

In [5]:
irs.columns

Index(['EIN', 'SEC_NAME', 'FRCD', 'SUBSECCD', 'TAXPER', 'ASSETS', 'INCOME',
       'NAME', 'ADDRESS', 'CITY', 'STATE', 'NTEECONF', 'NTEEFINAL', 'NAICS',
       'ZIP5', 'OUTNCCS', 'OUTREAS', 'RULEDATE', 'FIPS', 'FNDNCD', 'PMSA',
       'MSA_NECH', 'CASSETS', 'CFINSRC', 'CTAXPER', 'CTOTREV', 'ACCPER',
       'RANDNUM', 'NTEECC', 'NTEE1', 'LEVEL4', 'LEVEL1', 'NTMAJ10', 'MAJGRPB',
       'LEVEL3', 'LEVEL2', 'NTMAJ12', 'NTMAJ5', 'FILER', 'ZFILER'],
      dtype='object')

Another good function to check the dataframe is `irs.info(verbose=True)`. With it, we can see the object types of each column and how many are missing that particular information. This is important when we are choosing the columns to analyze. RIght now it is good to know that all observations have a value for CITY and ADDRESS since we want to separate the data for new york only. ZIP5 is missing only few, while STATE and FIPS are missing some thousands. They are still very usable, though.

In [6]:
irs.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1499450 entries, 0 to 1499449
Data columns (total 40 columns):
EIN          1499450 non-null int64
SEC_NAME     404909 non-null object
FRCD         1499450 non-null int64
SUBSECCD     1499450 non-null int64
TAXPER       1240366 non-null float64
ASSETS       1223112 non-null float64
INCOME       1223112 non-null float64
NAME         1499450 non-null object
ADDRESS      1499450 non-null object
CITY         1499450 non-null object
STATE        1498426 non-null object
NTEECONF     4518 non-null object
NTEEFINAL    1499450 non-null object
NAICS        1495191 non-null float64
ZIP5         1499353 non-null float64
OUTNCCS      1499450 non-null object
OUTREAS      3021 non-null object
RULEDATE     1499450 non-null int64
FIPS         1496643 non-null float64
FNDNCD       1499450 non-null int64
PMSA         506362 non-null float64
MSA_NECH     1175676 non-null float64
CASSETS      542418 non-null float64
CFINSRC      542418 non-null object
CTAXP

# Separating arts institutions

What we want to do now, is to create a subsection of the data, restricting the observation to art related and new york city based only. By doing that, we will have a much smaller sample to work with. For that, we need to make sense out of the many NTEE-related columns. Reading from the [dictionary of 2016](https://nccs-data.urban.org/dd2.php?close=1&form=BMF+08/2016) (unfortunately, we don't have a dictionary for 2018, but it should be the same), the column that makes sense to use is NTEECC - which stands for National Taxonomy of Exempt Entities Core Codes. This column is about the primary category of the institution. More about the NTEECC [here](https://nccs.urban.org/project/irs-activity-codes) and for the full list of codes, check [here](https://nccs.urban.org/publication/irs-activity-codes)


`.value_counts()` shows how many occurrances we had for each value in the NTEECC column. It is a good way to have a sense of the data in that column.

In [7]:
irs.NTEECC.value_counts()

X20     126651
X21     106577
T20      41652
W30      38862
S80      32814
B94      31545
B83      30338
T22      27565
N50      24052
J40      23871
Y41      21530
B82      19331
B11      18857
N60      18442
S41      18367
P20      17684
S20      17452
D20      17008
M24      16854
Y42      15969
X99      15623
B03      13899
A80      12418
A23      12399
S81      11610
N63      11508
W70      10709
B99      10470
A82      10291
O50       8823
         ...  
W057         1
P116         1
O195         1
H059         1
B055         1
L028         1
Q034         1
O024         1
G027         1
K014         1
J023         1
R113         1
F126         1
A197         1
C196         1
N037         1
C194         1
Q024         1
B126         1
D016         1
H128         1
G023         1
N194         1
L194         1
I113         1
I057         1
W055         1
L113         1
Q017         1
A055         1
Name: NTEECC, Length: 1443, dtype: int64

The "length: 1443" tell hos how many unique values we've got. If we only wanted that, we could have run `irs.NTEECC.nunique()` 

Now I want to create a subset of the data including only the ones with NTEECC starting with an A. The A code is for Arts, Culture & Humanities, and by definition is for "Private nonprofit organizations whose primary purpose is to promote appreciation for and enjoyment and understanding of the visual, performing, folk, and media arts; the humanities (archaeology, art history, modern and classical languages, philosophy, ethics, theology, and comparative religion); history and historical events; and/or communications (film, video, publishing, journalism, radio, television)."  [source and complete list of codes](https://nccs.urban.org/publication/irs-activity-codes)

We will probably have to have people sit, read all the codes and decide which ones are relevant. This will make our lives easier later, with less cases to operate with, and our maps less cluttered

In [8]:
As = irs[irs.NTEECC.str.startswith('A')]

The above was a quick command to create a new variable with the subset only for the A's. The way  to read it is: my new variable (As) will be the dataframe (irs) when the value for the column (irs.NTEECC) seen as a string (str) starts with A (startswith('A'))



Now we run `.value_counts()` again to see if it worked, and `.shape` to see the size of our subset

In [9]:
As['NTEECC'].value_counts()

A80     12418
A23     12399
A82     10291
A65      7900
A68      5756
A50      4992
A40      4537
A20      4520
A6B      4185
A62      3542
A6C      3100
A70      3031
A69      2425
A60      2236
A25      2193
A99      2013
A54      2005
A30      1901
A33      1881
A26      1867
A31      1510
A6E      1337
A61      1076
A03       977
A34       954
A84       952
A12       912
A32       879
A63       857
A116      736
        ...  
A022       33
A018       24
A192       23
A052       22
A193       19
A017       17
A123       17
A058       17
A056       16
A194       16
A125       13
A024       13
A195       13
A023        9
A053        9
A054        8
A025        8
A028        7
A199        6
A129        6
A057        6
A039        5
A029        5
A198        4
A015        4
A027        4
A019        2
A055        1
A197        1
A127        1
Name: NTEECC, Length: 98, dtype: int64

In [10]:
As.shape

(113138, 40)

# Separating New York City

## Option a: CITY column

Now I am trying to separate only the cases for New York City. There is indeed a city category, but in my experience, this is never something that works 100%
A quick check shows that there are a lot of unique cities, which makes is not viable to check one by one.

In [11]:
len(As.CITY.unique())

11138

A quick `value_counts()` so we can see this data

In [12]:
As.CITY.value_counts()

NEW YORK          3072
CHICAGO           1332
WASHINGTON        1279
LOS ANGELES       1080
HOUSTON            955
SAN FRANCISCO      948
BROOKLYN           937
PORTLAND           727
PHILADELPHIA       676
SEATTLE            642
AUSTIN             573
SAN DIEGO          553
BEACHWOOD          515
ATLANTA            508
DALLAS             489
SAINT LOUIS        455
MINNEAPOLIS        451
PITTSBURGH         435
COLUMBUS           424
DENVER             401
MADISON            363
CINCINNATI         353
SAN ANTONIO        352
BOSTON             352
MIAMI              347
COLUMBIA           341
BALTIMORE          340
HONOLULU           331
OAKLAND            327
INDIANAPOLIS       322
                  ... 
GRIMES               1
MELCHER DAL          1
ST JAMES CITY        1
ARDSLEY HDSN         1
LARCHWOOD            1
OGDEN DUNES          1
LAUREL HILL          1
BROUSSARD            1
MONON                1
MANGUM               1
TANGIER              1
ROMOLAND             1
KOELTZTOWN 

New York is the leading city, which makes sense. The problem is that it is not uncommon to see people putting their boroughs or even their neighbors as the city. So a quick check to see if the boroughs appear as city.

In [13]:
print('NEW YORK' in list(As.CITY))
print('QUEENS' in list(As.CITY))
print('BROOKLYN' in list(As.CITY))
print('STATEN ISLAND' in list(As.CITY))
print('MANHATTAN' in list(As.CITY))
print('BRONX' in list(As.CITY))

True
True
True
True
True
True


They all show up. It wouldn't be practical to search for other things such as neighborhoods and there is always the possibility of typos. For now we will just go with those options.

Now let's try to creat a subset using the column CITY with those names

In [14]:
nya = As[(As.CITY == 'NEW YORK') 
        | (As.CITY == 'QUEENS')
        | (As.CITY == 'BROOKLYN')
        | (As.CITY == 'STATEN ISLAND')
        | (As.CITY == 'MANHATTAN')
        | (As.CITY == 'BRONX')]

In [15]:
nya.CITY.value_counts()

NEW YORK         3072
BROOKLYN          937
BRONX             181
STATEN ISLAND     110
MANHATTAN          28
QUEENS              5
Name: CITY, dtype: int64

In [16]:
nya.shape

(4333, 40)

The value counts shows us how it was important to add the boroughs' names. If we hadn't, we would have missed more than a quarter of the institutions. 

## Option B: FIPS column

Check how many institutions we get using the fips code. If the number is bigger, we will just use it. If it is smaller, we check the value_counts of the cities to see what "cities" we are missing here.

In [17]:
As.FIPS.value_counts()

6037.0     3574
36061.0    3209
17031.0    2132
11001.0    1212
53033.0    1125
48201.0    1124
6073.0     1056
25017.0    1005
39035.0    1001
6075.0      968
6001.0      897
4013.0      883
6059.0      864
36047.0     840
48113.0     755
6085.0      747
27053.0     725
42101.0     702
42003.0     620
24031.0     610
12086.0     598
48453.0     567
41051.0     549
36081.0     545
13121.0     536
25025.0     504
36103.0     502
6067.0      485
51059.0     481
29189.0     478
           ... 
38091.0       1
38075.0       1
38065.0       1
38047.0       1
40055.0       1
40061.0       1
38045.0       1
20033.0       1
40069.0       1
38033.0       1
38031.0       1
38027.0       1
20049.0       1
40105.0       1
12077.0       1
38013.0       1
38011.0       1
8063.0        1
1011.0        1
40057.0       1
2230.0        1
46137.0       1
5087.0        1
46121.0       1
29063.0       1
27155.0       1
46115.0       1
22065.0       1
46111.0       1
47169.0       1
Name: FIPS, Length: 3055

Notice that these numbers all have a decimal point in the dataset. Need to remember this when creating the subset.

The county FIPS for the new york state is 36. The counties for New York City are:

    005 - Bronx
    047 - Kings (Brooklyn)
    061 - New York (Manhattan)
    081 - Queens
    085 - Richmond (Staten Island)



In [18]:
nyb = As[(As.FIPS == 36005.0) |
         (As.FIPS == 36047.0) |
         (As.FIPS == 36061.0) |
         (As.FIPS == 36081.0) |
         (As.FIPS == 36085.0)]

In [19]:
nyb.FIPS.value_counts()

36061.0    3209
36047.0     840
36081.0     545
36005.0     182
36085.0     120
Name: FIPS, dtype: int64

In [20]:
nyb.shape

(4896, 40)

We found more results in this method than in the other (4896 x 4333), but the number of brooklyn cases is smaller now (840 x 937), which suggests us that we are still missing cases. Let us make a quick test and see what namer for CITY are showing up:

In [21]:
nyb.CITY.value_counts()

NEW YORK            2981
BROOKLYN             843
BRONX                177
STATEN ISLAND        109
ASTORIA               78
FLUSHING              77
JAMAICA               42
LONG IS CITY          31
NEW YORK CITY         26
WOODSIDE              22
JACKSON HTS           22
FOREST HILLS          20
LONG ISLAND CITY      20
RIDGEWOOD             17
ELMHURST              14
SUNNYSIDE             13
BAYSIDE               13
EAST ELMHURST         10
NYC                    9
RICHMOND HILL          9
QUEENS VLG             9
JACKSON HEIGHTS        8
NEWYORK                8
FRESH MEADOWS          8
DOUGLASTON             7
COLLEGE POINT          7
LOS ANGELES            7
WASHINGTON             7
REGO PARK              7
ROSEDALE               6
                    ... 
PARLIN                 1
MONSEY                 1
ENGLEWOOD CLIFFS       1
LAGRANGEVILLE          1
HOLLYWOOD              1
SHORT HILLS            1
SCARSDALE              1
GENEVA                 1
BROOKIYN               1


This method is also not perfect. As we can see, we find institutions from baltimore, los angeles, miami... this shows us that there will still be cleaning to do later if we decide to go with this option

One way we could do it, is by eliminating the cases that don't have 'NY' in the STATE column. The problem is that we saw earlier that the column STATE was not as complete as FIPS, so we could end up losing valid cases. The way to do it would be:

In [22]:
nyb.STATE.value_counts()

NY    4768
NJ      29
CA      14
PA       8
FL       7
DC       7
CT       6
MA       5
GA       4
ME       3
MI       3
DE       3
MD       3
NC       3
NM       2
OR       2
IL       2
NV       2
TX       2
OK       2
MN       2
OH       2
KY       2
SC       2
WA       2
VI       1
RI       1
NH       1
VT       1
MS       1
HI       1
KS       1
LA       1
Name: STATE, dtype: int64

In [23]:
nyb2 = nyb[nyb.STATE == 'NY']

In [24]:
nyb2.CITY.value_counts()

NEW YORK                 2981
BROOKLYN                  843
BRONX                     177
STATEN ISLAND             109
ASTORIA                    78
FLUSHING                   77
JAMAICA                    42
LONG IS CITY               31
NEW YORK CITY              26
JACKSON HTS                22
WOODSIDE                   22
LONG ISLAND CITY           20
FOREST HILLS               20
RIDGEWOOD                  16
ELMHURST                   14
SUNNYSIDE                  13
BAYSIDE                    13
EAST ELMHURST              10
QUEENS VLG                  9
RICHMOND HILL               9
NYC                         9
NEWYORK                     8
FRESH MEADOWS               8
JACKSON HEIGHTS             8
DOUGLASTON                  7
REGO PARK                   7
COLLEGE POINT               7
ROSEDALE                    6
MANHATTAN                   5
QUEENS                      5
                         ... 
BROOKIYN                    1
BEACON                      1
NEPONSIT  

In [25]:
nyb2.shape 

(4768, 40)

We got rid of a little bit over a hundred cases, including some of those cities that were clearly wrong.

## Option C: ZIP5 column

I looked for a list of zip codes from new york city and copied here:

In [26]:
nyc_zipcodes = [10453, 10457, 10460,
10458, 10467, 10468,
10451, 10452, 10456,
10454, 10455, 10459, 10474,
10463, 10471,
10466, 10469, 10470, 10475,
10461, 10462,10464, 10465, 10472, 10473,
11212, 11213, 11216, 11233, 11238,
11209, 11214, 11228,
11204, 11218, 11219, 11230,
11234, 11236, 11239,
11223, 11224, 11229, 11235,
11201, 11205, 11215, 11217, 11231,
11203, 11210, 11225, 11226,
11207, 11208,
11211, 11222,
11220, 11232,
11206, 11221, 11237,
10026, 10027, 10030, 10037, 10039,
10001, 10011, 10018, 10019, 10020, 10036,
10029, 10035,
10010, 10016, 10017, 10022,
10012, 10013, 10014,
10004, 10005, 10006, 10007, 10038, 10280,
10002, 10003, 10009,
10021, 10028, 10044, 10065, 10075, 10128,
10023, 10024, 10025,
10031, 10032, 10033, 10034, 10040,
11361, 11362, 11363, 11364,
11354, 11355, 11356, 11357, 11358, 11359, 11360,
11365, 11366, 11367,
11412, 11423, 11432, 11433, 11434, 11435, 11436,
11101, 11102, 11103, 11104, 11105, 11106,
11374, 11375, 11379, 11385,
11691, 11692, 11693, 11694, 11695, 11697,
11004, 11005, 11411, 11413, 11422, 11426, 11427, 11428, 11429,
11414, 11415, 11416, 11417, 11418, 11419, 11420, 11421,
11368, 11369, 11370, 11372, 11373, 11377, 11378,
10302, 10303, 10310,
10306, 10307, 10308, 10309, 10312,
10301, 10304, 10305,
10314]


In [27]:
As.ZIP5.value_counts()

44122.0    422
20008.0    255
10003.0    200
10023.0    187
10025.0    162
10036.0    159
10001.0    154
10019.0    153
10013.0    142
10018.0    141
10011.0    136
94102.0    136
10022.0    132
10024.0    127
20036.0    124
94117.0    107
19107.0    106
60631.0    101
10016.0    100
10014.0     91
11201.0     90
0.0         90
94110.0     89
94103.0     86
11217.0     86
20006.0     82
10017.0     81
11215.0     81
10012.0     80
20001.0     78
          ... 
49653.0      1
56572.0      1
33539.0      1
78076.0      1
5774.0       1
62990.0      1
35207.0      1
30597.0      1
48647.0      1
38075.0      1
53556.0      1
73726.0      1
4410.0       1
99032.0      1
27824.0      1
58849.0      1
54457.0      1
13763.0      1
35214.0      1
38069.0      1
28219.0      1
43136.0      1
77968.0      1
28103.0      1
76135.0      1
70437.0      1
78368.0      1
76131.0      1
92513.0      1
77226.0      1
Name: ZIP5, Length: 22055, dtype: int64

The zipcodes are in float numbers, so I will make all make my list of zipcodes floats as well with a list comprehension. List comprehensions are one of the coolests tools python has. Make sure to read about them [here](https://www.pythonforbeginners.com/basics/list-comprehensions-in-python) but right now  as a quick way of creating/modifying a list based on conditions.

In [28]:
nyc_zipcodes = [float(zipcode) for zipcode in nyc_zipcodes]

In [29]:
nyc_zipcodes

[10453.0,
 10457.0,
 10460.0,
 10458.0,
 10467.0,
 10468.0,
 10451.0,
 10452.0,
 10456.0,
 10454.0,
 10455.0,
 10459.0,
 10474.0,
 10463.0,
 10471.0,
 10466.0,
 10469.0,
 10470.0,
 10475.0,
 10461.0,
 10462.0,
 10464.0,
 10465.0,
 10472.0,
 10473.0,
 11212.0,
 11213.0,
 11216.0,
 11233.0,
 11238.0,
 11209.0,
 11214.0,
 11228.0,
 11204.0,
 11218.0,
 11219.0,
 11230.0,
 11234.0,
 11236.0,
 11239.0,
 11223.0,
 11224.0,
 11229.0,
 11235.0,
 11201.0,
 11205.0,
 11215.0,
 11217.0,
 11231.0,
 11203.0,
 11210.0,
 11225.0,
 11226.0,
 11207.0,
 11208.0,
 11211.0,
 11222.0,
 11220.0,
 11232.0,
 11206.0,
 11221.0,
 11237.0,
 10026.0,
 10027.0,
 10030.0,
 10037.0,
 10039.0,
 10001.0,
 10011.0,
 10018.0,
 10019.0,
 10020.0,
 10036.0,
 10029.0,
 10035.0,
 10010.0,
 10016.0,
 10017.0,
 10022.0,
 10012.0,
 10013.0,
 10014.0,
 10004.0,
 10005.0,
 10006.0,
 10007.0,
 10038.0,
 10280.0,
 10002.0,
 10003.0,
 10009.0,
 10021.0,
 10028.0,
 10044.0,
 10065.0,
 10075.0,
 10128.0,
 10023.0,
 10024.0,
 10025.0,


It worked!
Now, let's create our subset with all case whose values on the ZIP5 column are in our nyc_zipcodes list, using the method `.isin()`

In [30]:
nyc = As[As.ZIP5.isin(nyc_zipcodes)]

In [31]:
nyc.CITY.value_counts()

NEW YORK            2744
BROOKLYN             810
BRONX                176
STATEN ISLAND        109
ASTORIA               78
FLUSHING              74
JAMAICA               38
LONG IS CITY          30
NEW YORK CITY         23
WOODSIDE              22
JACKSON HTS           22
FOREST HILLS          20
LONG ISLAND CITY      18
RIDGEWOOD             16
BAYSIDE               13
SUNNYSIDE             13
ELMHURST              12
EAST ELMHURST         10
QUEENS VLG             9
RICHMOND HILL          9
NYC                    9
JACKSON HEIGHTS        8
NEWYORK                8
FRESH MEADOWS          8
REGO PARK              7
DOUGLASTON             7
COLLEGE POINT          7
WASHINGTON             6
ROSEDALE               6
LOS ANGELES            6
                    ... 
UPR MONTCLAIR          1
PARLIN                 1
MONSEY                 1
LAGRANGEVILLE          1
HOLLYWOOD              1
SHORT HILLS            1
KAMUELA                1
PUTNAM VALLEY          1
HOBOKEN                1


In [32]:
nyc.shape # now we got rid of another three hundred cases

(4587, 40)

If we want to be thorough, we could try to find the cases that are in one option and not in the others, and analyze them individually to see if they should be in or out of our dataset. That will require some more time.

Another option is to choose one of the methods above, plot it by address+zip and take out all the cases that are outside of the city. This will be much quicker.

Of course, this would be done after we had finished selecting the NTEECC codes, which would make our sample smaller and easier to deal with.

If I had to choose right now between the options a, b or c, I would go with c, the one that works with zipcodes. CITY has too many typos, there are much more institutions without a FIPS value than a ZIP5 value. 

# Exporting as csv

Once our dataset is ready, we can use the `.to_csv()` method to export it

In [34]:
nyc.to_csv('nyc_arts.csv')

# Things to do / next steps

- choose the NTEECC codes
- after we are done with selecting the data, check the dictionary and choose what columns we want to work with
- generate a csv and share it with everyone